nav2_tasks/include/nav2_tasks/is_stuck_condition.hpp

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_TASKS__IS_STUCK_CONDITION_HPP
+#define NAV2_TASKS__IS_STUCK_CONDITION_HPP
+
+#include <string>
+#include <chrono>
+#include <ctime>
+#include <cmath>
+#include <thread>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs/msg/empty.hpp"
+#include "behavior_tree_core/condition_node.h"
+#include "nav2_robot/robot.hpp"
+
+namespace nav2_tasks
+{
+
+class IsStuckCondition : public BT::ConditionNode, public rclcpp::Node
+{
+
+public:
+  explicit IsStuckCondition(const std::string & condition_name)
+  : BT::ConditionNode(condition_name),
+    Node("IsStuckCondition")
+  {
+    RCLCPP_INFO(get_logger(), "IsStuckCondition::constructor");
+
+    vel_cmd_sub_ = this->create_subscription<geometry_msgs::msg::Twist>(
+      "cmd_vel",
+      [this](geometry_msgs::msg::Twist::SharedPtr msg) {current_vel_cmd_ = msg;});
+
+    odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
+      "odom",
+      [this](nav_msgs::msg::Odometry::SharedPtr msg) {current_velocity_ = msg;});
+  }
+
+  IsStuckCondition() = delete;
+
+  ~IsStuckCondition()
+  {
+  }
+
+  BT::NodeStatus tick() override
+  {
+    // Spin the node to get messages from the subscriptions
+    rclcpp::spin_some(this->get_node_base_interface());
+
+    if (isStuck()) {
+      logMessage("tick(): Robot stuck!");
+      return BT::NodeStatus::SUCCESS;
+    }
+
+    logMessage("tick(): Robot not stuck");
+    return BT::NodeStatus::FAILURE;
+  }
+
+  bool isStuck()
+  {
+    // TODO(orduno) Move algorithm to the robot class
+    // return robot_.isStuck();
+
+    using namespace std::chrono_literals;
+
+    if (current_velocity_ == nullptr) {
```

Comment:
Another benefit of using the robot class is that this kind of code isn't replicated in several places; there is one place that manages the pubs/subs.

Reply:
Agree, see comment above.